File: example/customersdotnet/Adapter.cs
using System;
using System.Threading.Tasks;
using MessagePack;
using NanoBus.Functions;

namespace Customers
{
    public class OutboundImpl : Outbound
    {
        private Invoker invoker;

        public OutboundImpl(Invoker invoker)
        {
            this.invoker = invoker;
        }

        public async Task SaveCustomer(Customer customer)
        {
            await invoker.Invoke("/customers.v1.Outbound/saveCustomer", customer);
        }

        public async Task CustomerCreated(Customer customer)
        {
            await invoker.Invoke("/customers.v1.Outbound/customerCreated", customer);
        }

        public async Task<Customer> FetchCustomer(ulong id)
        {
            var args = new GetCustomerArgs
            {
                Id = id,
            };
            return await invoker.InvokeWithReturn<Customer>("/customers.v1.Outbound/fetchCustomer", args);
        }
    }


    public delegate void Starter();

    public class Adapter
    {
        private HTTPHandlers handlers;
        private ICodec codec;
        private Invoker invoker;

        private static string GetEnvironmentVariable(string name, string defaultValue)
            => Environment.GetEnvironmentVariable(name) is string v && v.Length > 0 ? v : defaultValue;

        public Adapter()
        {
            var outboundBaseURL = GetEnvironmentVariable("OUTBOUND_BASE_URL", "http://localhost:32321/outbound");
            this.codec = new MsgPackCodec();
            var invoke = new HTTPInvoker(outboundBaseURL);
            this.invoker = new Invoker(invoke.Invoke, this.codec);
            this.handlers = new HTTPHandlers(this.codec);
        }

        public void Start()
        {
            var host = GetEnvironmentVariable("HOST", "localhost");
            var port = int.Parse(GetEnvironmentVariable("PORT", "9000"));
            this.handlers.Listen(port, host);
        }

        public void RegisterInboundHanders(Inbound handlers)
        {
            if (handlers.CreateCustomer != null)
            {
                this.handlers.RegisterHandler("/customers.v1.Inbound/createCustomer", async (input) =>
                {
                    var customer = this.codec.Decode<Customer>(input);
                    var result = await handlers.CreateCustomer(customer);
                    return codec.Encode(result);
                });
            }
            if (handlers.GetCustomer != null)
            {
                this.handlers.RegisterHandler("/customers.v1.Inbound/getCustomer", async (input) =>
                {
                    var args = this.codec.Decode<GetCustomerArgs>(input);
                    var result = await handlers.GetCustomer(args.Id);
                    return codec.Encode(result);
                });
            }
        }

        public Outbound NewOutbound()
        {
            return new OutboundImpl(this.invoker);
        }
    }
}

File: example/customersjava/src/main/java/io/nanobus/example/customers/adapter/Adapter.java
package io.nanobus.example.customers.adapter;

import io.nanobus.functions.Codec;
import io.nanobus.functions.HTTPHandlers;
import io.nanobus.functions.HTTPInvoker;
import io.nanobus.functions.Invoker;
import io.nanobus.functions.MsgPackCodec;

import io.nanobus.example.customers.models.*;
import io.nanobus.example.customers.service.*;

public class Adapter {
    static final int PORT = Integer.parseInt(System.getProperty("port", "9000"));
    static final String HOST = System.getProperty("host", "localhost");
    static final String OUTBOUND_BASE_URL = System.getProperty("outbound.base.url", "http://localhost:32321/outbound");

    private HTTPHandlers handlers;
    private Codec codec;
    private Invoker invoker;

    public Adapter() {
        this.codec = new MsgPackCodec();
        HTTPInvoker invoke = new HTTPInvoker(OUTBOUND_BASE_URL);
        this.invoker = new Invoker((namespace, operation, input) -> //
        invoke.invoke(namespace, operation, input), this.codec);
        this.handlers = new HTTPHandlers();
    }

    public void start() throws Exception {
        this.handlers.listen(PORT, HOST);
    }

    public void registerInboundHandlers(Inbound inbound) {
        handlers.registerHandler("customers.v1.Inbound", "createCustomer", input -> {
            Customer customer = codec.decode(input, Customer.class);
            return inbound.createCustomer(customer)
                .map(codec::encode);
        });
        handlers.registerHandler("customers.v1.Inbound", "getCustomer", input -> {
            GetCustomerArgs args = codec.decode(input, GetCustomerArgs.class);
            return inbound.getCustomer(args.getId())
                .map(codec::encode);
        });
    }

    public Outbound newOutbound() {
        return new OutboundImpl(invoker);
    }
}
